judge/tasks/user.py

```diff
@@ -20,16 +20,10 @@ def apply_submission_filter(queryset, options):
     if options['submission_results']:
         queryset = queryset.filter(result__in=options['submission_results'])
 
-    problem_code_glob_cache = {}
+    if options['submission_problem_glob'] != '*':
+        queryset = queryset.filter(problem__code__regex=fnmatch.translate(options['submission_problem_glob']))
```

Comment:
Might be interesting to experiment the subquery version to see if it offers any performance gains (perhaps also forcing a `STRAIGHT JOIN`):

```python
queryset = queryset.filter(problem__in=Problem.objects.filter(code__regex=fnmatch.translate(options['submission_problem_glob'])))
```

Also the MySQL regex flavour is not necessarily the same as that of python. Will this cause any issues?

Reply:
> Might be interesting to experiment the subquery version to see if it offers any performance gains (perhaps also forcing a `STRAIGHT JOIN`):
> 
> ```python
> queryset = queryset.filter(problem__in=Problem.objects.filter(code__regex=fnmatch.translate(options['submission_problem_glob'])))
> ```

With some testing on a table size of 100k submissions and 600 problems (unfortunately I don't have a database any larger):

```py
import fnmatch
import timeit

from judge.models import Problem, Submission
from judge.utils.raw_sql import use_straight_join

glob = r'a*'

def f():
    a = Submission.objects.select_related('problem', 'language', 'source').filter(problem__code__regex=fnmatch.translate(glob))
    list(a)

def g():
    a = Submission.objects.select_related('problem', 'language', 'source').filter(problem__in=Problem.objects.filter(code__regex=fnmatch.translate(glob)))
    list(a)

def h():
    a = Submission.objects.select_related('problem', 'language', 'source')
    use_straight_join(a)
    a = a.filter(problem__code__regex=fnmatch.translate(glob))
    list(a)

def i():
    a = Submission.objects.select_related('problem', 'language', 'source')
    use_straight_join(a)
    a = a.filter(problem__in=Problem.objects.filter(code__regex=fnmatch.translate(glob)))
    list(a)


for func in 'fghi':
    print(func, timeit.timeit(f'{func}()', globals=globals(), number=5))
```

```
f 11.460055768024176
g 10.058658519992605
h 10.914277172996663
i 10.024998323991895
```

It seems `g` or `i` is slightly faster (this has been reflected in the latest commit).

> Also the MySQL regex flavour is not necessarily the same as that of python. Will this cause any issues?

From testing with very weird strings, it seems that the subset of regex that is used by `fnmatch.translate` is also valid for MySQL. However, I can't find any online sources to confirm this. In particular, I'm unsure about the implications of the use of [re.escape](https://github.com/python/cpython/blob/master/Lib/fnmatch.py#L137) in `fnmatch.translate`. For now, I think it should be safe to use `fnmatch.translate` directly. If it ever does become an issue though, we could always restrict the character set to only `a-z0-9*?[]!`, since problem codes can't have any other characters anyways.